app/services/dqt_api_access.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require "net/http"
+require "jwt"
+
+class DqtApiAccess
+  class << self
+    def token
+      if @token.nil? || token_about_to_expire?
+        request = Net::HTTP::Get.new(uri)
+
+        data = {
+          grant_type: grant_type,
+          scope: scope,
+          client_id: client_id,
+          client_secret: client_secret,
+        }
+
+        request.set_form_data(data)
+
+        response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: use_ssl?, read_timeout: 20) do |http|
+          http.request(request)
+        end
+
+        if response.code == "200"
+          @token = JSON.parse(response.body)["access_token"]
+        else
+          raise "DQT access token could not be fetched"
+        end
+      end
+
+      @token
+    end
+
+  private
+
+    def token_about_to_expire?
+      expires_at = Time.zone.at(decoded_token.dig(0, "exp"))
+      Time.zone.now + token_expiry_leeway > expires_at
+    end
+
+    def decoded_token
+      JWT.decode @token, nil, false
```

Comment:
Why do we need to pass `nil, false` here?

Reply:
- `nil` is the key/password used for signing the token. in our case the token isn't signed hence `nil`
- `false` is whether we verify the signature. since the token isn't signed we can't verify it which is why it it set to false

Reply:
Does `JWT.decode @token` break badly?
```
  def decode(jwt, key = nil, verify = true, options = {}, &keyfinder)
    Decode.new(jwt, key, verify, DEFAULT_OPTIONS.merge(options), &keyfinder).decode_segments
  end
```
is the code for the method. It seems surprising that the default values would break. But if it does break, I guess passing two extra arguments is not the end of the world.

Reply:
oh i see your point, yeah looks like i can remove them then, the README says otherwise compared to the implementation

Reply:
ok it is needed as i just read the source correctly. default arguments in the code example you pasted are `nil, true` we need `nil, false` so therefore they are needed